students/DennisLee/lesson03/mailroom.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+
+# Initial donor list and the amounts they have donated
+donor_history = [
+            ['Red Herring', 65820.5, 31126.37, 15000],
+            ['Papa Smurf', 210.64, 1000],
+            ['Pat Panda', 55324.4],
+            ['Karl-Heinz Berthold', 3545.2, 10579.31],
+            ['Mama Murphy', 156316.99, 8500.3, 12054.33],
+            ['Daphne Dastardly', 82]
+        ]
+
+def manage_donors():
+    """
+    Display the menu of choices for donor management.
+
+    :return:  None.
+    """
+    choices = ("Send a Thank You", "Create a Report", "Quit")
+    while True:  # Infinite loop until user chooses to exit
+        # Print the menu list (with numbered choices)
+        print()
+        for i, j in enumerate(choices):
+            print(i + 1, j)
+
+        # Get the selection number
+        response = ''
+        while not response.isdigit() or int(
+                response) not in range(1, len(choices) + 1):
+            response = input("Type your selection: ").strip()
+        response = int(response)
+
+        # Call helper functions or exit
+        if response == 1:
+            send_thank_you()
+        elif response == 2:
+            create_a_report()
+        elif response == 3:
+            print("\nGoodbye.\n")
+            return
+        
+def send_thank_you():
+    """
+    Add new donations for new or existing donors, and send a thank-you
+    letter.
+
+    :return:  None.
+    """
+    donors = [x[0] for x in donor_history]  # Get list of donor names
+
+    # Get the donor name, show all donors, or quit
+    response = input(
+      "\nType the full donor name (or 'list' to show all donors, or 'quit'): "
+      ).strip()
+
+    if response.lower() == 'quit':
+        return
+
+    elif response.lower() == 'list':
+        print("\nLIST OF DONORS:", donors)
+        send_thank_you()  # Try getting a donor name again
+
+    else:
+        if response not in donors:  # Add to donor list if it's a new name
+            donors.append(response)
+            donor_history.append([response])
+
+        donation = '...'  # Seed donation variable to continue the while loop
+        # Make sure the donation amount is a valid, positive number
+        while donation.count('.') > 1 or donation.strip('0123456789.'
```

Comment:
It could make this section a bit more readable to just use an empty string for the donation initialization and then `while not donation`, or use some variation of `while True` with the number validation code in an `if` block, but that's just a suggestion.

Reply:
I had a lot of trouble with this section, as first I tried to use **.isnumeric()** and **.isdecimal()**, but those methods kept returning **False** whenever I tested a value with cents in it. The last part of the **while** statement was to ensure that **float(donation)** was positive, but that expression errored out if I set **donation** to `''`, so I gave in and initialized donation to a value that would flip the large Boolean expression to **True** on the first clause.

I wish the **while True** pathway had occurred to me. :( It would've saved me a headache to just exit the loop using a **break** statement by writing a conditional to look for the correct characteristics, instead of filtering out the incorrect stuff.

Reply:
This section is tricky to do well, especially with the toolkit available at this point in the course. We'll be adding exceptions to mailroom shortly which will help with validating for different forms of input. The `math` module of the Python Standard Library includes some methods that might be helpful to you here as well.